### PR TITLE
CRC files OK and documented

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

Transferred the libultra `crc.c` functions from OoT. I spent a couple of days learning about CRC, and wrote what I hope are fairly comprehensive/comprehensible notes on how it works and how it is implemented in libultra.

Also included @engineer124 's work on `voicecrc.c`, since it is almost identical to the second function in `crc.c`; it is used by the other voice functions, but requires no specialised structs or anything, so we decided it fit better in this PR.

I didn't know what to do with the following note that the OoT file had, so I have left it in for the time being:
```
// Valid addr up to 0x7FF
// It's the address of a block of 0x20 bytes in the mempak
// So that means the whole mempak has a 16-bit address space
```

Finally, this doesn't really fit here, but is not worth a PR by itself: I made a small change to calc_bss.py so that it can be used from the project root with tab-completion; I also chmodded it to be executable.

### CODE REVIEW

File: src/libultra/voice/voicecrc.c [hunk 1]
```diff
@@ -1,3 +1,48 @@
...
+/**
+ * This function is essentially the same as __osContDataCrc, but allows for a variable message length, specified by
+ * `numBytes`.
+ */
+u8 __osVoiceContDataCrc(u8* data, u32 numBytes) {
+    s32 ret = 0;
+    u32 bit;
+    u32 byte;
+
+    for (byte = numBytes; byte; byte--, data++) {
```
Comment (Thar0) [inline src/libultra/voice/voicecrc.c:21]: Is this `!= 0`?
```suggestion
    for (byte = numBytes; byte != 0; byte--, data++) {
```
We usually don't drop comparisons like this.

File: src/libultra/io/crc.c [hunk 2]
```diff
@@ -1,5 +1,161 @@
...
+ * Usually used as __osContAddressCrc(addr) | (addr << 5) to add the CRC to the end. The overall length of 10 + 5 bits
+ * allows the address + CRC to fit into one s16.
+ */
+u8 __osContAddressCrc(u16 addr) {
+    u32 addr32 = addr;
+    u32 ret = 0;
+    u32 bit;
+    s32 i;
+
+    /* ret is used as a shift register for the CRC */
```
Comment (Thar0) [inline src/libultra/io/crc.c:95]: We usually use `//` for comments inside function bodies

File: src/libultra/io/crc.c [hunk 3]
```diff
@@ -1,5 +1,161 @@
...
+// Valid addr up to 0x7FF
+// It's the address of a block of 0x20 bytes in the mempak
+// So that means the whole mempak has a 16-bit address space
+/**
+ * CRC-5 with the generating polynomial $x^5 + x^4 + x^2 + 1 $, AKA 0x15 = 0b(1)1 0101.
+ * It only works on the bits from 0x7FF = 11 11111111, i.e. 10 bits.
+ *
+ * Usually used as __osContAddressCrc(addr) | (addr << 5) to add the CRC to the end. The overall length of 10 + 5 bits
+ * allows the address + CRC to fit into one s16.
+ */
```
Comment (Thar0) [inline src/libultra/io/crc.c:88]: imo the oot docs should just be included for now, suggestion below, and revised later if a more appropriate location for them crops up.
```suggestion
/**
 * CRC-5 with the generating polynomial $x^5 + x^4 + x^2 + 1 $, AKA 0x15 = 0b(1)1 0101.
 * It only works on the bits from 0x7FF = 11 11111111, i.e. 10 bits.
 *
 * Usually used as __osContAddressCrc(addr) | (addr << 5) to add the CRC to the end. The overall length of 10 + 5 bits
 * allows the address + CRC to fit into one s16.
 *
 * `addr` is the address of a block in the mempak, only valid up to 0x7FF.
 */
```

**Review by Thar0 — CHANGES_REQUESTED** [review]
Looks good, I personally enjoyed the depth in the description of CRCs. Just some small things.

File: src/libultra/io/crc.c [hunk 4]
```diff
@@ -1,5 +1,160 @@
...
+ * `addr` is the address of a block in the mempak, only valid up to 0x7FF.
+ */
+u8 __osContAddressCrc(u16 addr) {
+    u32 addr32 = addr;
+    u32 ret = 0;
+    u32 bit;
+    s32 i;
+
+    // ret is used as a shift register for the CRC
+    for (bit = (1 << ADDRESS_CRC_MESSAGE_LENGTH); bit; bit >>= 1) {
```
Comment (AngheloAlf) [inline src/libultra/io/crc.c:95]:
```suggestion
    for (bit = (1 << ADDRESS_CRC_MESSAGE_LENGTH); bit != 0; bit >>= 1) {
```

File: src/libultra/voice/voicecrc.c [hunk 5]
```diff
@@ -1,3 +1,48 @@
...
+ * `numBytes`.
+ */
+u8 __osVoiceContDataCrc(u8* data, u32 numBytes) {
+    s32 ret = 0;
+    u32 bit;
+    u32 byte;
+
+    for (byte = numBytes; byte != 0; byte--, data++) {
+        // Loop over each bit in the byte starting with most significant
+        for (bit = (1 << (VOICE_CRC_LENGTH - 1)); bit; bit >>= 1) {
```
Comment (AngheloAlf) [inline src/libultra/voice/voicecrc.c:23]:
```suggestion
        for (bit = (1 << (VOICE_CRC_LENGTH - 1)); bit != 0; bit >>= 1) {
```

File: src/libultra/voice/voicecrc.c [hunk 6]
```diff
@@ -1,3 +1,48 @@
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/voicecrc/__osVoiceContDataCrc.s")
+#define VOICE_CRC_LENGTH 8
+#define VOICE_CRC_GENERATOR 0x85
+
+/**
+ * This function is essentially the same as __osContDataCrc, but allows for a variable message length, specified by
+ * `numBytes`.
+ */
+u8 __osVoiceContDataCrc(u8* data, u32 numBytes) {
```
Comment (Kenix3) [inline src/libultra/voice/voicecrc.c:16]:
```suggestion
u8 __osVoiceContDataCrc(u8* data, size_t numBytes) {
```

Is it possible to use size_t here? May not match.

**Review by Kenix3 — CHANGES_REQUESTED** [review]
Great job on this!

**Review by AngheloAlf — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Review by hensldm — APPROVED** [review]

**Pull request merged**